admin/cli/cmd/base/options.go

```diff
@@ -0,0 +1,161 @@
+	cmd.Flags().StringVar(&o.CertificateSecret, "ga-auth-cert-secret", o.CertificateSecret, "Keyvault secret that contains the Geneva Action auth certificate")
+	cmd.Flags().StringVar(&o.Endpoint, "admin-api-endpoint", o.Endpoint, "Admin API endpoint")
+	cmd.Flags().StringVar(&o.Host, "host", o.Host, "Host header to set in the request (useful for port-forward scenarios)")
+	cmd.Flags().BoolVar(&o.Insecure, "insecure-skip-verify", o.Insecure, "Skip TLS certificate verification")
+
+	// Mark required flags
+	for _, flag := range []string{
+		"ga-auth-cert-kv",
+		"ga-auth-cert-secret",
+		"ga-auth-tenant-id",
+		"ga-auth-client-id",
+		"ga-auth-scopes",
+		"admin-api-endpoint",
+	} {
+		if err := cmd.MarkFlagRequired(flag); err != nil {
+			return fmt.Errorf("failed to mark flag %q as required: %w", flag, err)
+		}
+	}
+
+	return nil
+}
+
+func (o *RawAuthOptions) Validate(ctx context.Context) (*ValidatedAuthTokenOptions, error) {
+	if o.KeyVaultURI == "" {
+		return nil, fmt.Errorf("ga-auth-cert-kv cannot be empty")
+	}
+
+	if o.CertificateSecret == "" {
+		return nil, fmt.Errorf("ga-auth-cert-secret cannot be empty")
+	}
+
+	if o.TenantID == "" {
+		return nil, fmt.Errorf("ga-auth-tenant-id cannot be empty")
+	}
+
+	if o.ClientID == "" {
+		return nil, fmt.Errorf("ga-auth-client-id cannot be empty")
+	}
+
+	if len(o.Scopes) == 0 {
+		return nil, fmt.Errorf("ga-auth-scopes cannot be empty")
+	}
+
+	if o.Endpoint == "" {
+		return nil, fmt.Errorf("admin-api-endpoint cannot be empty")
+	}
+
+	return &ValidatedAuthTokenOptions{
+		validatedAuthOptions: &validatedAuthOptions{
+			RawAuthOptions: o,
+		},
+	}, nil
+}
+
+type validatedAuthOptions struct {
+	*RawAuthOptions
+}
+
+type ValidatedAuthTokenOptions struct {
+	// Embed a private pointer that cannot be instantiated outside of this package.
+	*validatedAuthOptions
+}
+
+type completedAuthOptions struct {
+	Token    string
+	Endpoint string
+	Host     string
+	Insecure bool
+}
+
+type AuthOptions struct {
+	// Embed a private pointer that cannot be instantiated outside of this package.
+	*completedAuthOptions
+}
+
+func (o *validatedAuthOptions) Complete(ctx context.Context) (*AuthOptions, error) {
+
+	// Login with the certificate
+	token, err := auth.GetBearerToken(ctx, auth.CertificateAuthConfig{
+		KeyVaultURL:           o.KeyVaultURI,
+		CertificateSecretName: o.CertificateSecret,
+		TenantID:              o.TenantID,
+		ClientID:              o.ClientID,
+		Scopes:                o.Scopes,
+		SendX5C:               true,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get bearer token: %w", err)
+	}
+
+	return &AuthOptions{
+		completedAuthOptions: &completedAuthOptions{
+			Token:    token,
+			Endpoint: o.Endpoint,
+			Host:     o.Host,
+			Insecure: o.Insecure,
+		},
+	}, nil
+}
+
+func (o *AuthOptions) Execute(ctx context.Context) error {
+	logger := logr.FromContextOrDiscard(ctx)
```

Comment:
prefer to pass logger explicitly, if you must not, at least handle the error. implicit parameters in context + panic is a good way to end up in a bad place

Reply:
i handle the error. it is for sure better to fail than to silently drop logging and become blind when no logger was passed via the context